### PR TITLE
data instance for frame cloze problem

### PR DESCRIPTION
**Pull request by nikett**
This is similar to tuple instance.

### CODE REVIEW

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    def unpack_input(frame_as_string: str, kv_separator: str="\t"):
+        # no information is lost in lowercasing, and simplifies matching.
+        partialframe_query = frame_as_string.lower().split(kv_separator)
+        if len(partialframe_query) != 2:
+            raise RuntimeError("Unexpected number (not 2) of fields in frame: " + frame_as_string)
+        return {'content': partialframe_query[0], 'query': partialframe_query[1]}
+
+    @staticmethod
+    def given_slots_from(slots_csv: str, values_separator: str="###", kv_separator: str=":"):
+        return dict(map(lambda x: x.split(kv_separator), slots_csv.split(values_separator)))
```
Comment (DeNeutoy) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Could you add a comment describing this function? It's not clear immediately what it does. You could consider re-writing this like:
```
key_value_list = slots_csv.split(values_separator)
key_value_tuples = [pair.split(key_value_separator) for x in key_value_list]
key_value_dict = {key: value for key, value in key_value_tuples}

```

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                 # output is a phrase
+                 label: str=None):
+        super(FrameInstance, self).__init__(label)
+        self.text = text  # "event:plant absorb water###participant:water###agent:plant" TAB "agent:plant"
+
+    def __str__(self):
+        return 'FrameInstance( [' + ',\n'.join(self.text) + '] , ' + str(self.label) + ')'
+
+    @overrides
+    # accumulate words from each slot's phrase
```
Comment (nikett) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Done. I did it the java style, incorrectly.

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+QUES_SLOTVAL = "ques"  # this slot in the frame must be queried/completed.
+MAX_PHRASE_LEN = 6
+
+class FrameInstance(TextInstance):
+
+    """
+    A FrameInstance is a kind of TextInstance that has text in multiple slots. This generalizes a FrameInstance.
+    """
+    def __init__(self,
+                 # the input can have only few slots and not all and in any random order.
```
Comment (nikett) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Done.

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
+from typing import Dict, List
+
+import numpy
+from overrides import overrides
+
+from ..instance import TextInstance, IndexedInstance
+from ...data_indexer import DataIndexer
+
+
+# Functionality supported in this class.
```
Comment (nikett) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Done.

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        # e.g. from, participant:water, extract the expected slot value "water"
+        unpacked_input = cls.unpack_input(line)
+        query_slot = cls.slot_from(unpacked_input['query'])
+        given_slots = cls.given_slots_from(unpacked_input['content'])
+        padded_slots = cls.padded_slots_from(given_slots, query_slot['name'])
+        return cls(padded_slots, label=query_slot['val'])
+
+    @overrides
+    def to_indexed_instance(self, data_indexer: DataIndexer):
+        # a phrase in a slot, is converted from [list of words] to [list of wordids]
```
Comment (nikett) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Done.

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+    @staticmethod
+    # s: "participant:water" => name=participant, val=water
+    def slot_from(s: str, kv_separator: str=":"):
+        kv = s.split(kv_separator)
+        return {'name': kv[0], 'val': kv[1]}
+
+    @staticmethod
+    # s: "event:plant absorb water###participant:water" TAB "participant:water"
+    def unpack_input(s: str, kv_separator: str="\t"):
```
Comment (nikett) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Data processing in between the logic renders it unreadable. Small functions make the functionality easier to read, and debug.

File: tests/data/instances/text_classification/frame_instance_test.py [hunk 7]
```diff
@@ -1,47 +1,67 @@
...
+        self.padded_slots = ['plant', 'unk', 'unk', 'unk', 'unk',
+                             'plant absorb water', 'ques', 'unk', 'unk',
+                             'unk', 'unk', 'unk', 'unk', 'unk', 'unk',
+                             'unk', 'unk', 'unk', 'unk', 'unk', 'unk', 'unk',
+                             'unk', 'unk', 'unk', 'unk', 'water']
+        self.data_indexer = DataIndexer()
+        for word in ['plant', 'unk', 'absorb', 'ques', 'water', 'soil']:
+            self.data_indexer.add_word_to_index(word)
+
+    def tearDown(self):
```
Comment (DeNeutoy) [inline tests/data/instances/text_classification/frame_instance_test.py]: This is not required as you do nothing but call the super class.

File: tests/data/instances/text_classification/frame_instance_test.py [hunk 8]
```diff
@@ -1,47 +1,67 @@
...
+        super(TestFrameInstance, self).tearDown()
+
+    def test_convert_instance_to_indexed_instance(self):
+        instance = FrameInstance.read_from_line(self.line)
+        indexed_instance = instance.to_indexed_instance(self.data_indexer)
+        print(instance.label + "\t padded to: " + indexed_instance.label.__str__())
```
Comment (DeNeutoy) [inline tests/data/instances/text_classification/frame_instance_test.py]: Can you remove these print statements?

File: tests/data/instances/text_classification/frame_instance_test.py [hunk 9]
```diff
@@ -1,47 +1,67 @@
...
-    # TODO more test with boundary conditions on imperfect input.
     def test_slots_unwrap_correctly(self):
-        instance = FrameInstance.read_from_line(line)
+        instance = FrameInstance.read_from_line(self.line)
         # what we get
-        machine_label = "" + instance.label
-        machine_slotvals = ""+instance.text.__str__()
+        # TODO: consider when label is OOV or finalloc instead of finalloc:soil
+        machine_label = instance.label
+        machine_slotvals = instance.text
```
Comment (DeNeutoy) [inline tests/data/instances/text_classification/frame_instance_test.py]: No need to shorten variable names here - just use `machine_slot_values`.

File: tests/data/instances/text_classification/frame_instance_test.py [hunk 10]
```diff
@@ -1,47 +1,67 @@
...
+
+    def tearDown(self):
+        super(TestFrameInstance, self).tearDown()
+
+    def test_convert_instance_to_indexed_instance(self):
+        instance = FrameInstance.read_from_line(self.line)
+        indexed_instance = instance.to_indexed_instance(self.data_indexer)
+        print(instance.label + "\t padded to: " + indexed_instance.label.__str__())
+        print(indexed_instance.word_indices.__str__())
+        assert indexed_instance.label == [self.data_indexer.get_word_index('soil')]  # TODO
```
Comment (DeNeutoy) [inline tests/data/instances/text_classification/frame_instance_test.py]: remove this TODO.

File: tests/data/instances/text_classification/frame_instance_test.py (hunk 7 above)
Comment (nikett) [inline tests/data/instances/text_classification/frame_instance_test.py]: Done

File: tests/data/instances/text_classification/frame_instance_test.py (hunk 8 above)
Comment (nikett) [inline tests/data/instances/text_classification/frame_instance_test.py]: Done

File: tests/data/instances/text_classification/frame_instance_test.py (hunk 10 above)
Comment (nikett) [inline tests/data/instances/text_classification/frame_instance_test.py]: Done

File: tests/data/instances/text_classification/frame_instance_test.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    def test_slots_unwrap_correctly(self):
+        instance = FrameInstance.read_from_line(self.line)
+        # what we construct
+        machine_label = instance.label
+        machine_slot_values = instance.text
+        # what we expect
+        expected_label = "soil"
+        # do they match?
+        assert machine_label == expected_label
+        assert machine_slot_values == self.padded_slots
```
Comment (matt-gardner) [inline tests/data/instances/text_classification/frame_instance_test.py]: The label here is what you're trying to predict.  You want to have the model output "soil".  But it looks like there's nothing in this instance that tells the model what slot you're trying to query.  How would a model know which slot to fill, when it's trying to predict "soil"?

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    @classmethod
+    @overrides
+    def empty_instance(cls):
+        return IndexedFrameInstance([], label=None)
+
+    @overrides
+    def get_padding_lengths(self) -> Dict[str, int]:
+        # The parent class somehow understands the semantics of these dictionary keys.
+        # would be better if these keys were more explicit, e.g., as enumerations.
+        return {'num_sentence_words': MAX_PHRASE_LEN}
```
Comment (matt-gardner) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Yeah, there's no need to hard-code a length here, and that parameter is better specified in the model you build, anyway.  This will actually break the code we have that tries to be smart about letting you swap out word representations.  It looks like the tuple instance code that Mark linked to also will break the word representation stuff that we have.  Instead, what you should do is something like [this](https://github.com/allenai/deep_qa/blob/dd6f20cd0fde902cc54f58baff29f355a356c3cd/deep_qa/data/instances/entailment/sentence_pair_instance.py#L66-L71).

@DeNeutoy, I think we need way better documentation around how the whole data pipeline works.  I think it does some really nice things, but people have a hard time understanding it and how to use it correctly.

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+import numpy
+from overrides import overrides
+
+from ..instance import TextInstance, IndexedInstance
+from ...data_indexer import DataIndexer
+
+# TODO PR request for having these in the json as an application specific content
+# the slotnames can vary according to different end applications, e.g., a HowTo tuple, OpenIE tuple ...
+SLOTNAMES_ORDERED = ["agent", "beneficiary", "causer", "context", "definition", "event",
```
Comment (matt-gardner) [inline deep_qa/data/instances/text_classification/frame_instance.py]: If you want to always have a fixed set of slots that every dataset uses, where your input includes a one-hot vector for which slot you're trying to fill, or something like that, this should work.  The model's weights will be tied to particular slots, so a model trained on one slot configuration will not possibly generalize to any other slot configuration.

If you want to make this more modular, you could consider using a separate namespace in the data indexer to keep track of this for you, instead of hard-coding it here.  Something similar is done with tag labels [here](https://github.com/allenai/deep_qa/blob/dd6f20cd0fde902cc54f58baff29f355a356c3cd/deep_qa/data/instances/sequence_tagging/tagging_instance.py#L30).

File: deep_qa/data/instances/text_classification/frame_instance.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+MAX_PHRASE_LEN = 6
+
+
+class FrameInstance(TextInstance):
+
+    """
+    A FrameInstance is a kind of TextInstance that has text in multiple slots. This generalizes a FrameInstance.
+    """
+    def __init__(self,
+                 padded_slots: List[str],
```
Comment (nikett) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Fixed to sparse and dense frame, instead of unpadded and padded slots.

File: deep_qa/data/instances/text_classification/frame_instance.py (hunk 12 above)
Comment (nikett) [inline deep_qa/data/instances/text_classification/frame_instance.py]: I now understand. So something like this?

```  lengths = {}
for key in slot_word_lengths:
            lengths[key] = len(key)
return lengths```

File: deep_qa/data/instances/text_classification/frame_instance.py (hunk 13 above)
Comment (nikett) [inline deep_qa/data/instances/text_classification/frame_instance.py]: The reason to have this configurable was to be able to train any variant of the input with the same code, by passing a list with the config file upfront. This configuration allows ordering the input on the fly for any types of slots. 

It appears that there is no possibility to specify these list of slotnames in the config file, so, perhaps having this as a constant may be better than passing through a function. The data indexer does not really do any book keeping because the frame_instance takes care of it.

File: tests/data/instances/text_classification/frame_instance_test.py (hunk 11 above)
Comment (nikett) [inline tests/data/instances/text_classification/frame_instance_test.py]: Very good point. I was hoping to include a one-hot vector depicting the slot to fill. I thought that by specifying a special tag "ques", I could inform the model eventually. You are probably right, that in the data instance, my input could explicitly include a one-hot vector. I don't know how to do that so I would ask Mark.

File: deep_qa/data/instances/text_classification/frame_instance.py (hunk 13 above)
Comment (matt-gardner) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Oh, yeah, your preprocess scripts putting everything into this format will take care of that, I think.

Another consideration is that you'll be essentially wasting computation and modeling capacity on all of the un-filled slots by filling them with "unk".  You could either: (1) leave them empty, and let them be masked in whatever model computation you do later.  This will keep the model from wasting its capacity trying to figure out that "unk" means "unfilled".  Or (2) instead of having a fixed-sized input here, you could have a key-value input, where you give a list of (slot, value) pairs as input to the model.  In this setting, it's also a lot easier to think about how to pass in the query - it's just another (slot, value) pair, just with a `"QUERY"` value, or something similar.  Then instead of learning what is essentially an embedding of each slot type in the weights of a softmax layer, you can explicitly learn an embedding for them, and have that be part of your model.  Happy to talk about this more in person if it's hard to follow what I'm suggesting here.

File: deep_qa/data/instances/text_classification/frame_instance.py (hunk 12 above)
Comment (matt-gardner) [inline deep_qa/data/instances/text_classification/frame_instance.py]: Yeah, similar.  I would do something like this:

```python
all_slot_lengths = [self._get_word_sequence_lengths(slot_indices) for slot_indices in self.word_indices]
lengths = {}
for key in slot_lengths[0]:
    lengths[key] = max(slot_lengths[key] for slot_lengths in all_slot_lengths)
return lengths
```
